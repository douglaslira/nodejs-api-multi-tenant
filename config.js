require('dotenv').config()
var config = require('12factor-config')

var cfg = config({
  endUserBindPort: {
    env: 'END_USER_API_BIND_PORT',
    type: 'integer',
    default: 3002
  },
  adminBindPort: {
    env: 'ADMIN_API_BIND_PORT',
    type: 'integer',
    default: 3003
  },
  mongoHosts: {
    // comma delimited hostnames & ports, e.g. mongo1.blah.net:27017,mongo2.blah.net:27017,...
    env: 'MONGO_HOSTS',
    type: 'string',
    required: true
  },
  mongoReplicaSet: {
    env: 'MONGO_REPLICA_SET',
    type: 'string'
  },
  mongoUser: {
    env: 'MONGO_USER',
    type: 'string'
  },
  mongoPassword: {
    env: 'MONGO_PASSWORD',
    type: 'string'
  },
  mongoSSL: {
    env: 'MONGO_SSL',
    type: 'boolean',
    default: true
  },
  adminDomainName: {
    env: 'ADMIN_DOMAIN_NAME',
    type: 'string',
    default: '.app.localhost'
  },
  adminSessionExpiry: {
    env: 'ADMIN_SESSION_EXPIRY',
    type: 'integer',
    default: 60 * 60 * 8 // 8 hours
  },
  adminJwtSecret: {
    env: 'ADMIN_JWT_SECRET',
    type: 'string',
    default: 'c1a79517becd38f5339594bc11c358bc568eff96e8bda516c4717ac466660631' // node -e "console.log(require('crypto').randomBytes(32).toString('hex'));"
  },
  endUserSessionExpiry: {
    env: 'END_USER_SESSION_EXPIRY',
    type: 'integer',
    default: 60 * 60 * 24 * 90 // 3 months
  },
  endUserJwtSecret: {
    env: 'END_USER_JWT_SECRET',
    type: 'string',
    default: '142fddc6b2ecf2eab3d099cfa6310e1a987f20aee5a28812758ee5d713053abf' // node -e "console.log(require('crypto').randomBytes(32).toString('hex'));"
  },
  adminEmail: {
    env: 'ADMIN_EMAIL_ADDRESS',
    type: 'string',
    required: true,
    default: 'markbame.martires@gmail.com'
  },
  awsRegion: {
    // DONT SET THIS IN DEPLOYMENT - AUTO-DETECTED FROM ENV BY AMZN SDK
    env: 'AWS_REGION',
    type: 'string'
  },
  awsAccessKeyID: {
    // DONT SET THIS IN DEPLOYMENT - AUTO-DETECTED FROM ENV BY AMZN SDK
    env: 'AWS_ACCESS_KEY_ID',
    type: 'string'
  },
  awsSecretAccessKey: {
    // DONT SET THIS IN DEPLOYMENT - AUTO-DETECTED FROM ENV BY AMZN SDK
    env: 'AWS_SECRET_ACCESS_KEY',
    type: 'string'
  },
  smtpHost: {
    env: 'SMTP_HOST',
    type: 'string',
    default: 'email-smtp.eu-west-1.amazonaws.com'
  },
  smtpPort: {
    env: 'SMTP_PORT',
    type: 'integer',
    default: 587
  },
  smtpUsePool: {
    env: 'SMTP_USE_POOL',
    type: 'boolean',
    default: true
  },
  smtpUsername: {
    env: 'SMTP_USERNAME',
    type: 'string',
    default: 'AKIAI4FNMLWBJLB4U6SQ' // IAM user: newssocial-mailer
  },
  smtpPassword: {
    env: 'SMTP_PASSWORD',
    type: 'string',
    default: 'AsdScbGrbiPHRXCMOTIDc02AvckfJXtPvB3penAc+RmZ'
  }
})

module.exports = cfg
