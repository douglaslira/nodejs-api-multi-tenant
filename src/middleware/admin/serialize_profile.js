export default function(profile) {
  return {
    _id: profile._id,
    email: profile.email,
    firstname: profile.firstname,
    lastname: profile.lastname,
    registered: profile.registered
  }
}
