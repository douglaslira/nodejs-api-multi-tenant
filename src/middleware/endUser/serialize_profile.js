export default function(profile) {
  return {
    _id: profile._id,
    username: profile.username,
    email: profile.email,
    registered: profile.registered,
    verified: profile.verified,
    role: profile.role,
    avatar: profile.avatar,
    gender: profile.gender,
    firstname: profile.firstname,
    lastname: profile.lastname,
    bio: profile.bio
  }
}
