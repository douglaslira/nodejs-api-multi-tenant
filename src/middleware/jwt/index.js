import jwtMiddleware from './middleware'

export default jwtMiddleware
