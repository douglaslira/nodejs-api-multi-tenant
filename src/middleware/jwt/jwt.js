'use strict'

const jwt = require('jwt-simple')

export const now = () => new Date().getTime() / 1000

export const encoder = ({ secret, expireInSeconds, time = now }) => {
  return payload => {
    return jwt.encode(
      {
        exp: time() + expireInSeconds,
        ...payload
      },
      secret
    )
  }
}

export const decoder = ({ secret }) => {
  return encoded => {
    return jwt.decode(encoded, secret)
  }
}
