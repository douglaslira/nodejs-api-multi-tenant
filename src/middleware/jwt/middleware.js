'use strict'

import config from '../../../config'
import { decoder } from './jwt'

export const getCredentials = ctx => {
  if (!ctx.header || !ctx.header.authorization) {
    return
  }
  const parts = ctx.header.authorization.split(' ')
  if (parts.length === 2) {
    const scheme = parts[0]
    const credentials = parts[1]
    if (/^Bearer$/i.test(scheme)) {
      return credentials
    }
  }
}

export const decodeMiddleware = opts => {
  const decode = decoder(opts)
  return async (ctx, next) => {
    const credentials = getCredentials(ctx)
    if (credentials) {
      ctx.state.jwt = decode(credentials)
    }
    await next()
  }
}

const init = (app, secret) => {
  app.use(decodeMiddleware({ secret }))
}

export default init
