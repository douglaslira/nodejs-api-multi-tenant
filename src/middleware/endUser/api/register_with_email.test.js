'use strict'

import {
  register,
  confirmWithToken,
  confirmWithPin
} from './register_with_email'
import serialize_profile from '../serialize_profile'
import crypto from 'crypto'
import config from '../../../../config'
import { encoder } from '../../jwt/jwt'

jest.mock('../../../utils/runAsync')

describe('email registration', () => {
  let jwt, pin, ctx, enqueue, userRegister

  beforeEach(() => {
    userRegister = jest.fn().mockImplementation(user => ({
      ...user,
      _id: 'mock-id',
      role: 'test-role',
      registered: '2017-08-14'
    }))
    enqueue = jest.fn()
    // looks a bit like a jwt, but is stable over time :)
    jwt = jest
      .fn()
      .mockImplementation(payload =>
        crypto
          .createHash('sha256')
          .update(JSON.stringify(payload))
          .digest('hex')
      )
    pin = jest.fn().mockImplementation(user => 'abc12')
    ctx = {
      request: {
        body: {
          email: 'test@foo.bar',
          username: 'test-user',
          tags: ['test01', 'test02']
        }
      },
      state: {
        service: {
          domainName: 'test.foo.bar',
          mailer: {
            enqueue
          },
          User: {
            register: userRegister
          }
        }
      }
    }
  })

  it('returns serialized new-user as body when successful', async () => {
    const r = register(jwt, pin)
    await r(ctx)

    expect(ctx.body).toMatchSnapshot()
  })

  it('sets response status 201 (created) when successful', async () => {
    const r = register(jwt, pin)
    await r(ctx)

    expect(ctx.status).toBe(201)
  })

  it('enqueues confirm-registration email when successful', async () => {
    const r = register(jwt, pin)
    await r(ctx)

    expect(enqueue).toBeCalled()
    expect(enqueue.mock.calls[0][0]).toMatchSnapshot()
  })

  it('does not enqueue confirm-registration email when unsuccessful', async () => {
    userRegister.mockImplementation(() => {
      throw new Error('expected')
    })
    try {
      const r = register(jwt)
      await r(ctx)
      throw new Error('not expected - should not reach here')
    } catch (e) {
      expect(e.message).toBe('expected')
    }
    expect(enqueue).not.toBeCalled()
  })
})

describe('confirmWithToken', () => {
  let ctx, User, findOne, thrw, completeRegistration, login, activityStreams
  const token = encoder({
    secret: config.endUserJwtSecret,
    expireInSeconds: 60
  })({
    userId: 'abc123',
    email: 'foo@bar.com'
  })

  beforeEach(() => {
    findOne = jest.fn()
    thrw = jest.fn()
    login = jest.fn()
    activityStreams = {
      followTags: jest.fn()
    }
    completeRegistration = jest.fn()
    ctx = {
      throw: thrw,
      login
    }
    User = {
      findOne,
      completeRegistration
    }
  })

  it('looks up user matching id and email from token', async () => {
    await confirmWithToken(ctx, User, token, activityStreams)
    expect(findOne).toBeCalledWith({ _id: 'abc123', email: 'foo@bar.com' })
  })

  it('returns 400:no_such_user if user not found', async () => {
    findOne.mockImplementation(() => undefined)
    await confirmWithToken(ctx, User, token)

    expect(findOne).toBeCalledWith({ _id: 'abc123', email: 'foo@bar.com' })
    expect(thrw).toBeCalledWith(400, 'no_such_user')
    expect(login).not.toBeCalled()
    expect(activityStreams.followTags).not.toBeCalled()
  })

  it('returns 200 if valid token and user already verified', async () => {
    const user = {
      _id: 'abc123',
      email: 'foo@bar.com',
      verified: new Date()
    }
    findOne.mockImplementation(() => user)
    await confirmWithToken(ctx, User, token, 'my.domain.com', activityStreams)

    expect(findOne).toBeCalledWith({ _id: 'abc123', email: 'foo@bar.com' })
    expect(ctx.status).toBe(200)
    expect(ctx.body).toBeDefined()
    expect(login).toBeCalledWith({ ...user, domain: 'my.domain.com' })
    expect(activityStreams.followTags).not.toBeCalled()
  })

  it('returns 202 if valid token and user not previously verified', async () => {
    const verified = new Date()
    const user = { _id: 'abc123', email: 'foo@bar.com' }
    findOne.mockImplementation(() => user)
    completeRegistration.mockImplementation(u => ({
      ...u,
      verified
    }))
    await confirmWithToken(ctx, User, token, 'my.domain.com', activityStreams)

    expect(findOne).toBeCalledWith({ _id: 'abc123', email: 'foo@bar.com' })
    expect(ctx.status).toBe(202)
    expect(ctx.body).toBeDefined()
    expect(login).toBeCalled()
    expect(login.mock.calls[0][0]).toEqual({
      ...serialize_profile({ ...user, verified }),
      domain: 'my.domain.com'
    })
    expect(activityStreams.followTags).toBeCalled()
  })

  it('follows initial tags if user is verified successfully', async () => {
    const verified = new Date()
    const user = {
      _id: 'abc123',
      email: 'foo@bar.com',
      initial_tags: ['Foo', 'Bar']
    }
    findOne.mockImplementation(() => user)
    completeRegistration.mockImplementation(u => ({
      ...u,
      verified
    }))
    await confirmWithToken(ctx, User, token, 'my.domain.com', activityStreams)
    expect(activityStreams.followTags).toBeCalledWith(['Foo', 'Bar'], 'abc123')
  })
})

describe('confirmWithPin', () => {
  let ctx,
    User,
    pinFn,
    findById,
    thrw,
    completeRegistration,
    login,
    activityStreams

  beforeEach(() => {
    thrw = jest.fn()
    findById = jest.fn()
    pinFn = jest.fn()
    login = jest.fn()
    completeRegistration = jest.fn()
    activityStreams = {
      followTags: jest.fn()
    }
    User = {
      findById,
      completeRegistration
    }
    ctx = {
      login,
      throw: thrw
    }
  })

  it('looks up user with matching id', async () => {
    await confirmWithPin(ctx, User, pinFn, 'abc123', '1234')

    expect(findById).toBeCalledWith('abc123')
  })

  it('returns 400:no_such_user if user not found', async () => {
    findById.mockImplementation(() => undefined)
    await confirmWithPin(ctx, User, pinFn, 'abc123', '1234')

    expect(findById).toBeCalledWith('abc123')
    expect(thrw).toBeCalledWith(400, 'no_such_user')
    expect(login).not.toBeCalled()
    expect(activityStreams.followTags).not.toBeCalled()
  })

  it('returns 400:invalid_pin if pin doesnt match new pin created from user', async () => {
    findById.mockImplementation(() => ({ _id: 'abc123', email: 'foo@bar.com' }))
    pinFn.mockImplementation(() => 'valid-pin')
    await confirmWithPin(ctx, User, pinFn, 'abc123', 'invalid-pin')

    expect(findById).toBeCalledWith('abc123')
    expect(thrw).toBeCalledWith(400, 'invalid_pin')
    expect(login).not.toBeCalled()
    expect(activityStreams.followTags).not.toBeCalled()
  })

  it('returns 200 if valid pin but user already verified', async () => {
    const user = { _id: 'abc123', email: 'foo@bar.com', verified: new Date() }
    findById.mockImplementation(() => user)
    pinFn.mockImplementation(() => 'valid-pin')
    completeRegistration.mockImplementation(() => user)
    await confirmWithPin(
      ctx,
      User,
      pinFn,
      'abc123',
      'valid-pin',
      'my.domain.com',
      activityStreams
    )

    expect(findById).toBeCalledWith('abc123')
    expect(ctx.status).toBe(200)
    expect(ctx.body).toBeDefined()
    expect(login).toBeCalledWith({ ...user, domain: 'my.domain.com' })
    expect(activityStreams.followTags).not.toBeCalled()
  })

  it('invokes `completeRegistration` to set verified date of user if valid pin and not previously verified', async () => {
    const user = { _id: 'abc123', email: 'foo@bar.com' }
    findById.mockImplementation(() => user)
    pinFn.mockImplementation(() => 'valid-pin')
    completeRegistration.mockImplementation(() => user)
    await confirmWithPin(
      ctx,
      User,
      pinFn,
      'abc123',
      'valid-pin',
      'my.domain.com',
      activityStreams
    )

    expect(completeRegistration).toBeCalledWith(user)
    expect(login).toBeCalledWith({ ...user, domain: 'my.domain.com' })
  })

  it('returns 202 if valid pin and user not previously verified', async () => {
    const user = { _id: 'abc123', email: 'foo@bar.com' }
    findById.mockImplementation(() => user)
    pinFn.mockImplementation(() => 'valid-pin')
    completeRegistration.mockImplementation(() => user)
    await confirmWithPin(
      ctx,
      User,
      pinFn,
      'abc123',
      'valid-pin',
      'my.domain.com',
      activityStreams
    )

    expect(findById).toBeCalledWith('abc123')
    expect(ctx.status).toBe(202)
    expect(ctx.body).toBeDefined()
    expect(login).toBeCalledWith({ ...user, domain: 'my.domain.com' })
    expect(activityStreams.followTags).toBeCalled()
  })

  it('follows initial tags if user is verified successfully', async () => {
    const user = {
      _id: 'abc123',
      email: 'foo@bar.com',
      initial_tags: ['Foo', 'Bar']
    }
    findById.mockImplementation(() => user)
    pinFn.mockImplementation(() => 'valid-pin')
    completeRegistration.mockImplementation(() => user)
    await confirmWithPin(
      ctx,
      User,
      pinFn,
      'abc123',
      'valid-pin',
      'my.domain.com',
      activityStreams
    )
    expect(activityStreams.followTags).toBeCalledWith(['Foo', 'Bar'], 'abc123')
  })
})
