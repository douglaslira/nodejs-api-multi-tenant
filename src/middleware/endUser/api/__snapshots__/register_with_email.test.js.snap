// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email registration enqueues confirm-registration email when successful 1`] = `
Object {
  "from": "test.foo.bar <no-reply@foo.bar>",
  "language": "en",
  "payload": Object {
    "confirmationUrl": "https://test.foo.bar/confirm-email-with-token/fb5f5852c8006a90e3eb9321ab840dbe15f2f92118fac2ea3b2718e99a70cca0",
    "domainName": "test.foo.bar",
    "pin": "abc12",
    "user": Object {
      "_id": "mock-id",
      "avatar": undefined,
      "bio": undefined,
      "email": "test@foo.bar",
      "firstname": undefined,
      "gender": undefined,
      "lastname": undefined,
      "registered": "2017-08-14",
      "role": "test-role",
      "username": "test-user",
      "verified": undefined,
    },
  },
  "to": "test@foo.bar",
  "type": "confirm-registration",
}
`;

exports[`email registration returns serialized new-user as body when successful 1`] = `
Object {
  "_id": "mock-id",
  "avatar": undefined,
  "bio": undefined,
  "email": "test@foo.bar",
  "firstname": undefined,
  "gender": undefined,
  "lastname": undefined,
  "registered": "2017-08-14",
  "role": "test-role",
  "username": "test-user",
  "verified": undefined,
}
`;
